kernel/waker_generic/src/lib.rs

```diff
@@ -0,0 +1,114 @@
+//! (e.g., the `task` crate) directly, meaning it is usable in the `task` crate.
+//! This works by accepts generic actions (closures) for waking the [Waker](core::task::Waker)
+//! and blocking the [Blocker].
+
+#![no_std]
+
+extern crate alloc;
+
+use alloc::sync::Arc;
+// @tsoutman: as I said before, I think we definitely do need a preemption-safe mutex here.
+//            It's not technically required, but it'd prevent task switch thrashing
+//            between the task that is being blocked and the task that is waking the waker.
+use spin::Mutex; 
+
+/// Creates a new waker and blocker pair that are associated with each other.
+///
+/// # Arguments
+/// * `wake_action`: the action that will be taken by the waker when it is ready
+///    to wake up the blocker, i.e., allow it to stop blocking.
+///
+/// The blocker can be used to block a task until the waker is woken.
+pub fn new_waker<W>(wake_action: W) -> (core::task::Waker, Blocker)
+where
+    W: Fn() + Send + Sync + 'static, // required by the bounds on core::task::Waker::from
+{
+    let woken = Arc::new(Mutex::new(false));
+    (
+        core::task::Waker::from(Arc::new(Waker {
+            woken: woken.clone(),
+            wake_action,
+        })),
+        Blocker { woken },
+    )
+}
+
+/// A blocker that blocks in a loop until the associated waker is woken.
+///
+/// To obtain a `Blocker` and its associated [`core::task::Waker`], call [`new_waker()`].
+pub struct Blocker {
+    /// Whether the waker has been woken.
+    ///
+    /// This is shared with the Waker associated with this Blocker.
+    /// If true, this Blocker can cease blocking and return to its caller.
+    woken: Arc<Mutex<bool>>,
+}
+impl Blocker {
+    /// Performs the given `block_action` within this `Blocker`,
+    /// looping until the associated waker is woken.
+    ///
+    /// If the waker was already woken prior to this function being called,
+    /// it will return immediately.
+    ///
+    /// # Arguments
+    /// * `block_action`: a closure that is called in a loop until its woker is woken.
+    ///    The return value `R` of the closure is any arbitrary value that will be held until
+    ///    *after* the blocker has released the lock on its inner state shared with the waker.
+    ///    * This can be used, for example, to yield the CPU to another task.
+    pub fn block<B, R>(&self, block_action: B)
+    where
+        B: Fn() -> R,
+    {
+        // Care must be taken not to introduce race conditions.
+        // After creating the waker, the wake condition must be checked to ensure
+        // that it did not get woken prior to registering the waker;
+        // otherwise, the waker will never be woken, and this function will loop forever.
+        loop {
+            let woken = self.woken.lock();
+            if *woken {
+                // *woken = false;  // To @tsoutsman: why do this? there's no point, right?
```

Comment:
Wakers can be reused. If `woken` isn't reset, subsequent calls to `block` will immediately return.

Reply:
alright, i will restore that statement. 

In the current code i didn't see any examples of wakers being reused. What would the interface to that look like?
I only ask because i wouldn't want to be overly restrictive with the interface we expose to access a waker, e.g., requiring the waker to be consumed.

On a related note -- should we consider making waker `wake()` itself when it gets dropped? I might be overthinking it, but if another caller sets a task's registered waker instance, the existing waker instance (if any) will be overwritten and dropped. If we assume that wakers can never be spuriously woken, then this wouldn't be a good idea. But it's not ideal to have a waker get registered with a task and then it not actually ever be woken up. 
Of course we can just remove that interface for now.... just wondering if you'd need it in `dreadnought`.

Reply:
E.g. An async I/O driver that wakes the waker every time new data is available.

> I might be overthinking it, but if another caller sets a task's registered waker instance, the existing waker instance (if any) will be overwritten and dropped

Currently, it can only be done through a `JoinableTaskRef` of which only one exists.

> If we assume that wakers can never be spuriously woken

In the context of async, wakers can be woken spuriously, but it leads to the executor unnecessarily polling the future again.

> just wondering if you'd need it in `dreadnought`

I don't think so. If a future doesn't wake the waker, then it shouldn't be woken.

As a side note, we aren't using a `generic_waker::Waker`, we are using a `core::task::Waker` (a glorified vtable) so the drop handler would actually be called in `core::task::Waker::from` in `new_waker`.

Reply:
Agree on all counts. Thanks!